File: crypto/src/test/cplusplus/multilinearextension.cpp
/*
 * Copyright (c) 2024-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <boost/test/unit_test.hpp>
#include <ranges>

#include "circuitbuilder.h"
#include "customizableconstraintsystem.h"
#include "hypercube.h"
#include "matrixdense.h"
#include "multilinearextension.h"
#include "point.h"
#include "solinas62.h"
#include "solinas62extension.h"
#include "vectordense.h"
#include "util.h"

using namespace blacknet::crypto;

BOOST_AUTO_TEST_SUITE(MultilinearExtensions)

using E = Solinas62Ring;
using EE = Solinas62RingDegree2;

BOOST_AUTO_TEST_CASE(meta) {
    MultilinearExtension mle{E(1), E(2), E(3), E(4), E(5), E(6), E(7), E(8)};
    BOOST_TEST(1 == mle.degree());
    BOOST_TEST(3 == mle.variables());
}

BOOST_AUTO_TEST_CASE(add) {
    MultilinearExtension a{E(1), E(2), E(3), E(4)};
    MultilinearExtension b{E(5), E(6), E(7), E(8)};
    MultilinearExtension c{E(6), E(8), E(10), E(12)};
    BOOST_TEST(c == a + b);
    BOOST_TEST(c == b + a);
}

BOOST_AUTO_TEST_CASE(mul) {
    MultilinearExtension a{E(1), E(2), E(3), E(4)};
    E b(3);
    MultilinearExtension c{E(3), E(6), E(9), E(12)};
    BOOST_TEST(c == a * b);
}

BOOST_AUTO_TEST_CASE(sub) {
    MultilinearExtension a{E(99), E(98), E(97), E(96)};
    E b(3);
    MultilinearExtension c{E(96), E(95), E(94), E(93)};
    BOOST_TEST(c == a - b);
}

BOOST_AUTO_TEST_CASE(bind) {
    MultilinearExtension a{E(1), E(2), E(3), E(4), E(5), E(6), E(7), E(8)};
    MultilinearExtension b{E(1), E(2), E(3), E(4)};
    MultilinearExtension c{E(3), E(4)};
    MultilinearExtension d{E(4)};

    MultilinearExtension mle(a);
    mle.bind(E(0));
    BOOST_TEST(b == mle);
    mle.bind(E(1));
    BOOST_TEST(c == mle);
    mle.bind(E(1));
    BOOST_TEST(d == mle);

    std::vector<E> evaluations(4);
    mle = a;
    mle.bind(E(-2));
    a.bind<E(-2), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(-1));
    a.bind<E(-1), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(0));
    a.bind<E(0), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(1));
    a.bind<E(1), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(2));
    a.bind<E(2), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(3));
    a.bind<E(3), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
    mle = a;
    mle.bind(E(4));
    a.bind<E(4), util::Assign<E>>(evaluations);
    BOOST_TEST(mle() == evaluations);
}

BOOST_AUTO_TEST_CASE(matrix) {
    Hypercube<E> hc(3);
    MatrixDense<E> a(2, 4, {
        E(30), E(31), E(32), E(33),
        E(43), E(44), E(45), E(46),
    });
    MultilinearExtension mle(a);
    for (const auto& [pair, b] : std::views::zip(
            std::ranges::subrange(hc.splittedBegin(2, 4), hc.splittedEnd()),
            std::ranges::subrange(hc.decomposedBegin(), hc.decomposedEnd())
        )) {
        const auto& [row, column] = pair;
        BOOST_TEST((a[row, column] == mle(b)));
    };
}

BOOST_AUTO_TEST_CASE(polynomial) {
    using P = Solinas62RingDegree4;
    Hypercube<E> hc(2);
    P a{E(71), E(72), E(73), E(74)};
    MultilinearExtension<E> mle(a);
    for (const auto& [index, b] : std::views::zip(
            std::ranges::subrange(hc.composedBegin(), hc.composedEnd()),
            std::ranges::subrange(hc.decomposedBegin(), hc.decomposedEnd())
        )) {
        BOOST_TEST(a.coefficients[index] == mle(b));
    };
}

BOOST_AUTO_TEST_CASE(vector) {
    Hypercube<E> hc(3);
    VectorDense<E> a{E(63), E(64), E(65), E(66), E(67), E(68), E(69), E(70)};
    MultilinearExtension mle(a);
    for (const auto& [index, b] : std::views::zip(
            std::ranges::subrange(hc.composedBegin(), hc.composedEnd()),
            std::ranges::subrange(hc.decomposedBegin(), hc.decomposedEnd())
        )) {
        BOOST_TEST(a[index] == mle(b));
    };
}

BOOST_AUTO_TEST_CASE(ringvector) {
    Hypercube<E> hc(3);
    VectorDense<EE> a{
        EE{75, 76},
        EE{77, 78},
        EE{78, 79},
        EE{79, 80},
    };
    MultilinearExtension<E> mle(a);
    for (const auto& [pair, b] : std::views::zip(
            std::ranges::subrange(hc.splittedBegin(4, 2), hc.splittedEnd()),
            std::ranges::subrange(hc.decomposedBegin(), hc.decomposedEnd())
        )) {
        const auto& [row, column] = pair;
        BOOST_TEST(a.elements[row].coefficients[column] == mle(b));
    };
}

BOOST_AUTO_TEST_CASE(circuit) {
    MultilinearExtension<E> mle({E(2), E(3), E(5), E(7)});
    Point<E> x{E(11), E(13)};

    using Builder = CircuitBuilder<E, 2>;
    Builder circuit;
    using Circuit = MultilinearExtension<E>::Circuit<Builder>;
    Circuit mle_circuit(&circuit, Builder::Variable::Type::Input, 2);
    using PointCircuit = Point<E>::Circuit<Builder>;
    PointCircuit x_circuit(&circuit, Builder::Variable::Type::Input, 2);
    mle_circuit(x_circuit);

    CustomizableConstraintSystem<E> ccs(circuit.ccs());
    VectorDense<E> z = ccs.assigment();
    std::ranges::copy(mle.coefficients, std::back_inserter(z.elements));
    std::ranges::copy(x, std::back_inserter(z.elements));

    using Assigner = MultilinearExtension<E>::Assigner<Builder::degree()>;
    Assigner mle_assigner(mle, &z.elements);
    BOOST_TEST(mle(x) == mle_assigner(x));
    BOOST_TEST(ccs.isSatisfied(z));
}

BOOST_AUTO_TEST_SUITE_END()

File: crypto/src/test/rust/multilinearextension.rs
/*
 * Copyright (c) 2024-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_crypto::freemodule::FreeModule;
use blacknet_crypto::hypercube::Hypercube;
use blacknet_crypto::matrix::{DenseMatrix, DenseVector};
use blacknet_crypto::multilinearextension::MultilinearExtension;
use blacknet_crypto::operation::Double;
use blacknet_crypto::polynomial::Polynomial;
use core::iter::zip;

type R = blacknet_crypto::pervushin::PervushinField;

#[test]
fn meta() {
    let mle = MultilinearExtension::from([1, 2, 3, 4, 5, 6, 7, 8].map(R::from));
    assert_eq!(mle.degree(), 1);
    assert_eq!(mle.variables(), 3);
}

#[test]
fn add() {
    let a = MultilinearExtension::from([1, 2, 3, 4].map(R::from));
    let b = MultilinearExtension::from([5, 6, 1, 5].map(R::from));
    let c = MultilinearExtension::from([6, 8, 4, 9].map(R::from));
    assert_eq!(a + b, c);
}

#[test]
fn dbl() {
    let a = MultilinearExtension::from([1, 2, 3, 4].map(R::from));
    let b = MultilinearExtension::from([2, 4, 6, 8].map(R::from));
    assert_eq!(a.double(), b);
}

#[test]
fn sub() {
    let a = MultilinearExtension::from([99, 98, 97, 96].map(R::from));
    let b = R::from(3);
    let c = MultilinearExtension::from([96, 95, 94, 93].map(R::from));
    assert_eq!(a - b, c);
}

#[test]
fn neg() {
    let a = MultilinearExtension::from([7, 0, -1, 0].map(R::from));
    let b = MultilinearExtension::from([-7, 0, 1, 0].map(R::from));
    assert_eq!(-a, b);
}

#[test]
fn mul() {
    let a = MultilinearExtension::from([1, 2, 3, 4].map(R::from));
    let b = R::from(2);
    let c = MultilinearExtension::from([2, 4, 6, 8].map(R::from));
    assert_eq!(a * b, c);
}

#[test]
fn bind() {
    let a = MultilinearExtension::from([1, 2, 3, 4, 5, 6, 7, 8].map(R::from));
    let b = MultilinearExtension::from([1, 2, 3, 4].map(R::from));
    let c = MultilinearExtension::from([3, 4].map(R::from));
    let d = MultilinearExtension::from([4].map(R::from));

    let mut mle = a.clone();
    mle.bind(R::from(0));
    assert_eq!(mle, b);
    mle.bind(R::from(1));
    assert_eq!(mle, c);
    mle.bind(R::from(1));
    assert_eq!(mle, d);

    mle = a.clone();
    mle.bind(R::from(-2));
    let evaluations = a.hypercube_with_var::<-2>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(-1));
    let evaluations = a.hypercube_with_var::<-1>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(0));
    let evaluations = a.hypercube_with_var::<0>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(1));
    let evaluations = a.hypercube_with_var::<1>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(2));
    let evaluations = a.hypercube_with_var::<2>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(3));
    let evaluations = a.hypercube_with_var::<3>();
    assert_eq!(evaluations.elements(), mle.hypercube());
    mle = a.clone();
    mle.bind(R::from(4));
    let evaluations = a.hypercube_with_var::<4>();
    assert_eq!(evaluations.elements(), mle.hypercube());
}

#[test]
fn matrix() {
    let hc = Hypercube::<R>::new(3);
    #[rustfmt::skip]
    let a = DenseMatrix::new(2, 4, [
        30, 31, 32, 33,
        43, 44, 45, 46,
    ].map(R::from).into());
    let mle = MultilinearExtension::from(a.clone());
    zip(hc.iter_rank2(2, 4), hc.iter_vertex())
        .for_each(|(index, b)| assert_eq!(mle.point(&b), a[index]));
}

#[test]
fn vector() {
    let hc = Hypercube::<R>::new(3);
    let a = DenseVector::from([63, 64, 65, 66, 67, 68, 69, 70].map(R::from));
    let mle = MultilinearExtension::from(a.clone());
    zip(hc.iter_index(), hc.iter_vertex())
        .for_each(|(index, b)| assert_eq!(mle.point(&b), a[index]));
}

#[test]
fn module() {
    let hc = Hypercube::<R>::new(2);
    let a = FreeModule::from([71, 72, 73, 74].map(R::from));
    let mle = MultilinearExtension::from_iter(a);
    zip(hc.iter_index(), hc.iter_vertex())
        .for_each(|(index, b)| assert_eq!(mle.point(&b), a[index]));
}
